### PR TITLE
Command-line argument parsing for coqdoc via Arg

### PR DESCRIPTION
**Pull request by artagnon**
In an effort to neatly separate out argument parsing, minimize the number of globals, and create an automated help output, migrate coqdoc to use the OCaml stdlib Arg to do command-line argument parsing. 

I've tried to follow the example of configure, and picked a binary for which ~~users are unlikely to have scripts~~ there is a lot of automated testing.

```text
coqdoc [options] <input file>...
Available options are:
  --html               Produce a HTML document (default)
  --latex              Produce a LaTeX document
  --texmacs            Produce a TeXmacs document
  --raw                Produce a text document
  --dvi                Output the DVI
  --ps                 Output the PostScript
  --pdf                Output the Pdf
  --stdout             Write output to stdout
  -o <file>            Write output to file <file>
  --output <file>      Write output to file <file>
  -d <dir>             Output files into directory <dir>
  --directory <dir>    Output files into directory <dir>
  -g                   Skip proofs (gallina)
  --gallina            Skip proofs
  -s                   No titles for files (short)
  --short              No titles for files
  -l                   Light mode (only defs and statements)
  --light              Light mode (only defs and statements)
  -t <string>          Give a title to the document
  --title <string>     Give a title to the document
  --body-only          Suppress LaTeX/HTML header and trailer
  --no-preamble        Suppress LaTeX/HTML header and trailer
  --with-header <file> Prepend <file> as html reader
  --with-footer <file> append <file> as html footer
  --no-index           Do not output the index
  --multi-index        Index split in multiple files
  --index <string>     Set index name (default is index)
  --toc                Output a table of contents
  --table-of-contents  Output a table of contents
  --vernac-file <file> consider <file> as a .v file
  --tex-file <file>    Consider <file> as a .tex file
  -p <string>          Insert <string> in LaTeX preamble
  --preamble <string>  Insert <string> in LaTeX preamble
  --files-from <file>  Read file names to process in <file>
  --files <file>       Read file names to process in <file>
  --glob-from <file>   Read globalization information from <file>
  --no-glob            Don't use any globalization information (no links will be inserted at identifiers)
  -q                   Quiet mode (default)
  --quiet              Quiet mode (default)
  --verbose            Verbose mode
  --no-externals       No links to Coq standard library
  --external <url>+<d> set URL for external library <d>
  --coqlib_url <url>   Set URL for Coq standard library (default: http://coq.inria.fr/distrib/V8.16+alpha/stdlib/)
  --coqlib <dir>       Set the path where Coq files are installed
  -R <dir>+<coqdir>    map physical dir to Coq dir
  -Q <dir>+<coqdir>    Map physical dir to Coq dir
  --latin1             Set ISO-8859-1 mode
  --utf8               Set UTF-8 mode
  --charset <string>   Set HTML charset
  --inputenc <string>  Set LaTeX input encoding
  --interpolate        Try to typeset identifiers in comments using definitions in the same module
  --raw-comments       Raw comments
  --parse-comments     Parse regular comments
  --plain-comments     Consider comments as non-literate text
  --toc-depth <int>    Don't include TOC entries for sections below level <int>
  --no-lib-name        Don't display "Library" before library names in the toc
  --lib-name <string>  Call top level toc entries <string> instead of "Library"
  --lib-subtitles      First line comments of the form (** * ModuleName : text *) will be interpreted as subtitles
  --inline-notmono     Use a proportional width font for inline code (possibly with a different color)
  --version            Display coqdoc version
  -help                Display this list of options
  --help               Display this list of options
```

### CODE REVIEW

**Review by Alizter — COMMENTED** [review]
A good first try, but it looks like you missed some options. For example `-q` as an alias for `--quiet` is missing and in general coqdoc seems to take both `--` and `-`. This seems to be something Arg is not capable of.

It seems like implementing our own argument parser is the better option than using existing libraries. We want to have finer control over how Coq parses things since we have some quirky parsing conventions.

**Comment by silene** [comment]
> and picked a binary for which users are unlikely to have scripts.

That is a incredibly optimistic statement. I never run Coqdoc by hand, it is always run through a script, because the actual command is several lines long. It is not just a matter of typing just `coqdoc`. For example, Flocq's script invoking Coqdoc has not changed for the last ten years. And I am sure some people have scripts for Coqdoc that date back even further. In fact, CoqMakefile's users presumably never even ran Coqdoc by hand and always through a script. So, at the very least, you need to fix `tools/CoqMakefile.in`, so that they are still able to generate documentation.

**Comment by artagnon** [comment]
> For now, let's add the aliases as duplicate entries.

I did something a bit smarter ;)

> > and picked a binary for which users are unlikely to have scripts.
> 
> That is a incredibly optimistic statement. I never run Coqdoc by hand, it is always run through a script, because the actual command is several lines long. It is not just a matter of typing just `coqdoc`. For example, Flocq's script invoking Coqdoc has not changed for the last ten years. And I am sure some people have scripts for Coqdoc that date back even further. In fact, CoqMakefile's users presumably never even ran Coqdoc by hand and always through a script. So, at the very least, you need to fix `tools/CoqMakefile.in`, so that they are still able to generate documentation.

Oh, I see. I've added the missing and deprecated options. Hopefully, the CI will not be _filled_ with failures :)

**Comment by artagnon** [comment]
Hm, what does this CI failure mean?

```text
File "tools/coqdoc/cmdArgs.ml", line 72, characters 17-29:
Error: Unbound constructor Arg.Rest_all
```

Isn't Arg part of OCaml stdlib?

**Comment by silene** [comment]
> Isn't Arg part of OCaml stdlib?

Sure, but `Arg.Rest_all` is a very recent addition to the standard library, so it will not be available for most users.

**Comment by artagnon** [comment]
> Sure, but `Arg.Rest_all` is a very recent addition to the standard library, so it will not be available for most users.

Could you advise on what I could use instead? None of the other keywords seem appropriate.

**Comment by gares** [comment]
About CI, we don't build the doc of te projecs we test, even if there is some. Eg math-comp has a few scripts built around coqdoc which are not tested by Coq's CI.

**Comment by artagnon** [comment]
Do you suggest asking the CI to build these too?

The upside of targeting coqdoc for this patch: there's pretty good CI coverage of the command-line options, and we don't need additional tests.

**Comment by gares** [comment]
no, I'd rather not break compatibility.

**Comment by maximedenes** [comment]
The description of this PR is missing a few words on the motivation for the patch.

**Comment by Alizter** [comment]
@maximedenes 

The motivation for this patch was discussed on Zulip, I will summarise it here. Various Coq binaries have their own routines for parsing arguments and generating help pages/man pages. This has lead to more harder-to-maintain code.

In order to automatically generate help and man pages it would make sense for all Coq binaries to use the same argument parsing mechanism. As a proof-of-concept, coqdoc has been selected here.

It is the eventual goal to consolidate all argument parsing code into a common library (or external library).

File: tools/coqdoc/cmdArgs.ml [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+let add_input_files f = prefs := { !prefs with files = what_file f :: !prefs.files }
+let usage_msg = "coqdoc [options] <input file>...\nAvailable options are:"
+
+let parse_args () =
+(* Deprecated options *)
+let deprecated_opts =
+  ["-html"; "-latex"; "-texmacs"; "-raw"; "-dvi"; "-ps"; "-pdf"; "-stdout"; "-output"; "-directory"; "-gallina"; "-short"; "-light"; "-title"; "-body-only"; "-with-header"; "-with-footer"; "-no-index"; "-multi-index"; "-index"; "-toc"; "-table-of-contents"; "-vernac-file"; "-tex-file"; "-preamble"; "-files-from"; "-glob-from"; "-no-glob"; "-quiet"; "-verbose"; "-no-externals"; "-external"; "-coqlib_url"; "-coqlib"; "-latin1"; "-utf8"; "-charset"; "-inputenc"; "-interpolate"; "-raw-comments"; "-parse-comments"; "-plain-comments"; "-toc-depth"; "-no-lib-name"; "-lib-name"; "-lib-subtitles"; "-inline-notmono"; "-version"] in
+  let new_argv = Array.map (fun s -> match List.find_opt (fun m -> m = s) deprecated_opts with
+    | Some _ -> deprecation_warn s; Printf.sprintf "-%s" s
+    | None -> s) Sys.argv in
```
Comment (Alizter) [inline tools/coqdoc/cmdArgs.ml:234]: I suggested that in the future we would perhaps deprecate single `-` arguments, however no such decision has been made as of yet. Please don't do this in this patch.

**Review by Alizter — CHANGES_REQUESTED** [review]
I would prefer to do this in a way that does not change how (or which) command line arguments are parsed.

**Comment by Alizter** [comment]
Using the `Arg` library has several shortcomings. The fact that we can only register a single command line option for each action means that we have to duplicate the entries. This obviously makes the help documentation terrible.

The other solution might be to preprocess `-` to become `--` for all options but this also has some disadvantages. It presupposes that we have made a decision to use standard linux convention (which the dev team has not), and it would mean the help screen would display things like `--R` and `--Q` which is unhelpful and confusing.

The middle ground might be to preprocess all `-` to become `--` only for long options. This is still supposing we want to go down this route, but I think for coqdoc it makes sense.

**Comment by artagnon** [comment]
All done; removed the deprecation message. I've dropped `-noindex`, `-nopreamble`, `-noexternals` in favor of the `-no-...` versions (both used to work earlier), and dropped `-V` in favor of `-version` (both used to work earlier), for the sake of not bloating up the `--help` output or making the code harder to maintain. The tests pass, and I hope existing users with scripts aren't inconvenienced too much -- they were relying on undocumented features.

**Review by Alizter — DISMISSED** [review]
Let's not hope for convenience. I would rather that all the existing command line options, including the undocumented ones were supported. I don't think it is a large price to pay to pollute the help screen a bit.

Silently dropping commands is, in my opinion, a recipe for disaster and user disgruntlement. We can and ought to deprecate commands which we no longer wish to support, but that should be for a future PR. The fact that `coqdoc -v` is no longer supported the kind of oversight this way of doing things gives rise to.

The test-suite is rather light when it comes to coqdoc, and the CI doesn't test many documentation jobs that exist alongside those projects. This is a blind spot and I don't think we should rely on it.

In the end I believe this PR will give a more principles way of parsing command line arguments but I don't think it will be the end of the story.

**Comment by Alizter** [comment]
@artagnon Sorry it's been a while. Whats the update with the latest push?

**Comment by artagnon** [comment]
Hi @Alizter, it's all fixed in the latest push. Please verify.

File: tools/coqdoc/cpretty.mll [hunk 2]
```diff
@@ -26,9 +27,9 @@
     let n = String.length s in
     let rec count c i =
       if i == n then c,i else match s.[i] with
-	| '\t' -> count (c + (8 - (c mod 8))) (i + 1)
-	| ' ' -> count (c + 1) (i + 1)
-	| _ -> c,i
+  | '\t' -> count (c + (8 - (c mod 8))) (i + 1)
```
Comment (Alizter) [inline tools/coqdoc/cpretty.mll:30]: What is going on here with the indentation?

Comment (artagnon) [reply tools/coqdoc/cpretty.mll]: Fixed.

Comment (Alizter) [reply tools/coqdoc/cpretty.mll]: I should have been more clear. The entire file has strange indentation changes.

Comment (artagnon) [reply tools/coqdoc/cpretty.mll]: All fixed, modulo tabs versus spaces.

**Comment by artagnon** [comment]
Okay, so the problem arose in the first place because `cpretty.mll` was using a mix of tabs and spaces for indentation, and VSCode was unhappy with this: there was probably some bug in the auto-conversion to all-space indentation, and I think I've fixed all the instances now by hand, and even corrected indentation in a few places where it was incorrect.

**Review by Alizter — APPROVED** [review]
I had a few detailed looks at this PR and the functionality appears to be preserved. Thanks a lot @artagnon for the time and attention spent.

**Comment by Alizter** [comment]
@coqbot run full ci

**Comment by Alizter** [comment]
I will run the full ci just in case. Note that we don't test coqdoc at all in the CI.

**Comment by Alizter** [comment]
I also wouldn't mind any comments from @coq/coqdoc-maintainers.

**Comment by SkySkimmer** [comment]
>I will run the full ci just in case. Note that we don't test coqdoc at all in the CI.

WDYM? There are some tests in the test suite.

**Comment by Alizter** [comment]
@coqbot merge now

**Pull request merged**